src/librustc/ty/layout.rs

```diff
@@ -511,41 +512,70 @@ pub struct Struct {
     /// If true, the size is exact, otherwise it's only a lower bound.
     pub sized: bool,
 
-    /// Offsets for the first byte of each field.
+    /// Offsets for the first byte of each field, ordered in increasing order.
+    /// The ith element of this vector is not always the ith field of the struct.
     /// FIXME(eddyb) use small vector optimization for the common case.
     pub offsets: Vec<Size>,
```

Comment:
To have the increasing order `offsets` (instead of declaration order - is it really necessary?) you need to rename the field to `permuted_offsets` or `ordered_offsets` and (re-)add a `field_offset` method.